lib/xapi-stdext-date/date.ml

```diff
@@ -37,19 +37,37 @@ type iso8601 = Ptime.date * Ptime.time * print_type
 let of_dt print_type dt = let (date, time) = dt in (date, time, print_type)
 let to_dt (date, time, _) = (date, time)
 
-let of_string x =
+let best_effort_iso8601_to_rfc3339 x =
+  (* (a) add dashes
+   * (b) add UTC tz if no tz provided *)
   let x =
     try
-      (* if x doesn't contain dashes, insert them, so that ptime can parse x *)
-      Scanf.sscanf x "%04d%02d%02dT%s" (fun y mon d rest ->
-        Printf.sprintf "%04d-%02d-%02dT%s" y mon d rest
-      )
-    with _ -> x
+      Scanf.sscanf x "%04d%02d%02dT%s"
+        (fun y mon d rest ->
+          Printf.sprintf "%04d-%02d-%02dT%s" y mon d rest)
+    with _ ->
+      x
+  in
+  let tz =
+    try
+      Scanf.sscanf x "%04d-%02d-%02dT%02d:%02d:%02d%s"
+        (fun _ _ _ _ _ _ tz -> Some tz)
+    with _ -> None
   in
-  match x |> Ptime.of_rfc3339 |> Ptime.rfc3339_error_to_msg with
-  | Error (`Msg e) -> invalid_arg (Printf.sprintf "date.ml:of_string: %s" e)
+  match tz with
+  | None | Some "" ->
+    (* the caller didn't specify a tz. we must try to add one so that ptime can at least attempt to parse *)
+    (Printf.sprintf "%sZ" x, PrintLocal)
```

Comment:
Was localtime picked previously? It'd seem picking UTC would make it more stable.

Reply:
It failed completely previously (i.e. `of_string "2020-12-12T08:08:08"` threw an error)

I don't think choosing UTC when no timezone is specified is right. The library is working on the assumption that no timezone means localtime - for example:

```
localtime () |> to_string = "20201211T14:41:56"
```

If we assume that no timezone means UTC in `of_string` then we'd get

```
localtime () |> to_string |> of_string |> to_string = "20201211T14:41:56Z"
```

which is of course true only if the localtime is UTC

Reply:
I added a test for this ^